File: ai_houkai/timeparse.py
"""Lenient timestamp parsing shared by the CLI, MCP server and HTTP API.

`recall(since=…, until=…)` takes Unix timestamps (floats).  Humans and other
tools prefer ISO dates or relative spans, so the user-facing layers funnel
their inputs through :func:`parse_timestamp` to get a float the store accepts.
"""

from __future__ import annotations

import re
import time
from datetime import datetime, timezone

_REL_RE = re.compile(r"^\s*(\d+(?:\.\d+)?)\s*([smhdw])\s*$", re.IGNORECASE)
_REL_SECONDS = {"s": 1, "m": 60, "h": 3600, "d": 86_400, "w": 604_800}


def parse_timestamp(value: object, *, now: float | None = None) -> float | None:
    """Coerce *value* into a Unix timestamp (seconds since epoch).

    Accepts, in order:
      • ``None`` / empty string → ``None`` (no bound)
      • ``int`` / ``float`` → taken as an epoch timestamp verbatim
      • a relative span like ``"7d"``, ``"24h"``, ``"30m"`` → ``now`` minus that
        span (handy for "since 7 days ago")
      • a bare numeric string → parsed as an epoch timestamp
      • an ISO-8601 date/datetime (``"2026-06-14"`` or
        ``"2026-06-14T10:30:00"``) → its epoch timestamp; a trailing ``Z`` is
        honoured and naive values are interpreted as UTC

    Raises :class:`ValueError` for anything else so callers can surface a clear
    message instead of silently dropping the filter.
    """
    if value is None:
        return None
    if isinstance(value, bool):  # bool is an int subclass — reject explicitly
        raise ValueError(f"invalid timestamp: {value!r}")
    if isinstance(value, (int, float)):
        return float(value)

    text = str(value).strip()
    if not text:
        return None

    rel = _REL_RE.match(text)
    if rel:
        amount = float(rel.group(1))
        unit = rel.group(2).lower()
        base = time.time() if now is None else now
        return base - amount * _REL_SECONDS[unit]

    try:
        return float(text)
    except ValueError:
        pass

    iso = text[:-1] + "+00:00" if text.endswith("Z") else text
    try:
        dt = datetime.fromisoformat(iso)
    except ValueError as exc:
        raise ValueError(
            f"invalid timestamp {text!r}: expected epoch seconds, an ISO-8601 "
            f"date/datetime, or a relative span like '7d'"
        ) from exc
    if dt.tzinfo is None:
        dt = dt.replace(tzinfo=timezone.utc)
    return dt.timestamp()

File: go/internal/timeparse/timeparse.go
// Package timeparse provides lenient timestamp parsing shared by the CLI, MCP
// server and HTTP API.
//
// Recall's since/until filters take Unix timestamps (seconds). Humans and other
// tools prefer ISO dates or relative spans, so the user-facing layers funnel
// their inputs through Parse to get a value the store accepts.
package timeparse

import (
	"fmt"
	"regexp"
	"strconv"
	"strings"
	"time"
)

var relRE = regexp.MustCompile(`(?i)^\s*(\d+(?:\.\d+)?)\s*([smhdw])\s*$`)

var relSeconds = map[string]float64{
	"s": 1,
	"m": 60,
	"h": 3600,
	"d": 86400,
	"w": 604800,
}

// isoLayouts are tried in order for non-relative, non-numeric input. Naive
// values (no zone) are interpreted as UTC, mirroring the Python implementation.
var isoLayouts = []string{
	time.RFC3339Nano,
	time.RFC3339,
	"2006-01-02T15:04:05",
	"2006-01-02T15:04",
	"2006-01-02 15:04:05",
	"2006-01-02",
}

// Parse coerces value into a Unix timestamp (seconds since epoch).
//
// It accepts, in order:
//   - an empty string → (0, false, nil): no bound
//   - a relative span like "7d", "24h", "30m" → now minus that span
//   - a bare numeric string → parsed as an epoch timestamp verbatim
//   - an ISO-8601 date/datetime (e.g. "2026-06-14" or "2026-06-14T10:30:00");
//     a trailing "Z" is honoured and naive values are interpreted as UTC
//
// The ok result reports whether a bound was set (false for empty input).
// A non-nil error is returned for anything that cannot be parsed, so callers
// can surface a clear message instead of silently dropping the filter.
func Parse(value string) (ts float64, ok bool, err error) {
	return ParseAt(value, time.Now())
}

// ParseAt is Parse with an explicit "now" for relative spans (testable).
func ParseAt(value string, now time.Time) (float64, bool, error) {
	text := strings.TrimSpace(value)
	if text == "" {
		return 0, false, nil
	}

	if m := relRE.FindStringSubmatch(text); m != nil {
		amount, _ := strconv.ParseFloat(m[1], 64)
		unit := strings.ToLower(m[2])
		return float64(now.Unix()) - amount*relSeconds[unit], true, nil
	}

	if n, perr := strconv.ParseFloat(text, 64); perr == nil {
		return n, true, nil
	}

	iso := text
	if strings.HasSuffix(iso, "Z") {
		iso = iso[:len(iso)-1] + "+00:00"
	}
	for _, layout := range isoLayouts {
		if t, perr := time.Parse(layout, iso); perr == nil {
			return float64(t.Unix()), true, nil
		}
		// Naive layouts (no zone) parse in UTC via ParseInLocation.
		if t, perr := time.ParseInLocation(layout, text, time.UTC); perr == nil {
			return float64(t.Unix()), true, nil
		}
	}

	return 0, false, fmt.Errorf(
		"invalid timestamp %q: expected epoch seconds, an ISO-8601 "+
			"date/datetime, or a relative span like '7d'", text)
}
